### PR TITLE
use a direct connection to the sqlite db, when available

### PR DESCRIPTION
**Pull request by adnelson**
@sfoulkes

This should improve the performance of the repo a lot; using SQLite directly means we don't need a subprocess or O(n) iteration through the nix store to check paths, so operations become a lot faster.

### CODE REVIEW

File: src/servenix/server/servenix.py [hunk 1]
```diff
@@ -6,6 +6,16 @@
 import re
 from subprocess import check_output, Popen, PIPE, CalledProcessError
 from threading import Thread
+# Special-case here to address a runtime bug I've encountered
```
Comment (sfoulkes) [inline src/servenix/server/servenix.py]: Maybe you should outline what the bug is?

Comment (adnelson) [inline src/servenix/server/servenix.py]: I think the error message that it throws does that job... no?

File: src/servenix/server/servenix.py [hunk 2]
```diff
@@ -55,6 +65,21 @@ def __init__(self, nix_store_path, nix_bin_path, nix_state_path,
...
+        # using nix-store. This is much faster, but is unavailable on
+        # some systems.
+        try:
+            db_path = join(nix_state_path, "nix", "db", "db.sqlite")
+            query = "select * from ValidPaths limit 1"
+            db_con = sqlite3.connect(db_path)
+            db_con.execute(query).fetchall()
+            # If this succeeds, assign the db_con attribute.
+            self._db_con = db_con
+        except Exception as err:
```
Comment (sfoulkes) [inline src/servenix/server/servenix.py:84]: Catch a more specific error?

Comment (adnelson) [inline src/servenix/server/servenix.py:84]: I considered that but I'm worried about it being too specific; essentially if any step within that try block fails then we can't connect to the database and we want to bomb out.

File: src/servenix/server/servenix.py [hunk 3]
```diff
@@ -87,24 +112,37 @@ def store_path_from_hash(self, store_object_hash):
...
-                # valid path according to nix-store.
-                if not self.check_in_store(path):
-                    break
+        if self._db_con is not None:
+            # If we have a direct database connection, use this to check
+            # path existence.
+            query = ("select path from ValidPaths where path like '{}%'"
+                     .format(join(self._nix_store_path, store_object_hash)))
+            with self._db_con:
+                paths = self._db_con.execute(query).fetchall()
```
Comment (sfoulkes) [inline src/servenix/server/servenix.py]: You should be using bind variables:

```
query = "select path from ValidPaths where path like %s"
with self._db_con:
    paths = self._db_con.execute(query, join(self._nix_store_path, store_object_hash)).fetchall()
```

File: src/servenix/server/servenix.py [hunk 4]
```diff
@@ -124,15 +162,29 @@ def check_in_store(self, store_path):
...
+        # If it is on the filesystem, it doesn't necessarily mean
+        # that it's a registered path in the store. Check that
+        # here.
+        # If we have a connection to the database, all we have to
+        # do is look in the database.
+        if self._db_con is not None:
+            query = ("select path from ValidPaths where path = '{}'"
+                     .format(store_path))
+            with self._db_con:
+                results = self._db_con.execute(query).fetchall()
```
Comment (sfoulkes) [inline src/servenix/server/servenix.py]: bind variables

File: src/servenix/server/servenix.py (hunk 2 above)
Comment (sfoulkes) [inline src/servenix/server/servenix.py:84]: Fair enough.

File: src/servenix/server/servenix.py (hunk 1 above)
Comment (sfoulkes) [inline src/servenix/server/servenix.py]: I guess.

File: src/servenix/server/servenix.py (hunk 3 above)
Comment (sfoulkes) [inline src/servenix/server/servenix.py]: Alright, working example is something like this.  The wildcard character needs to be in the value that you're subbing in:

```
query = "select path from ValidPaths where path like %s"
wildcard_path = join(self._nix_store_path, store_object_hash) + "%"
with self._db_con:
    paths = self._db_con.execute(query, wildcard_path).fetchall()
```

**Pull request merged**